### PR TITLE
#638 Zip a list of files (or a Directory)

### PR DESCRIPTION
**Pull request by SharplEr**
For #638 

**Problem**
There aren't any OOP way to create ZIP archive from `Directory`.

**Solution**

Add `Zip implements Input` which wraps the `Directory` to `Input` as result of work with `ZipOutputStream`.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#740](https://github.com/yegor256/cactoos/issues/740) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#740](https://github.com/yegor256/cactoos/issues/740) is assigned to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx), here is [why](https://www.0crat.com/footprint/C63314D6Z/742211992). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by 0crat** [comment]
@fabriciofx[/z](https://www.0crat.com/u/fabriciofx) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by 0crat** [comment]
@fabriciofx[/z](https://www.0crat.com/u/fabriciofx) this job was assigned to you 8 days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by 0crat** [comment]
The user @fabriciofx[/z](https://www.0crat.com/u/fabriciofx) resigned from [#740](https://github.com/yegor256/cactoos/issues/740), please stop working. Reason for job resignation: It is older than 10 days, see [§8](http://www.zerocracy.com/policy.html#8)

**Comment by 0crat** [comment]
Resigned on delay, see [§8](http://www.zerocracy.com/policy.html#8): -15 points just awarded to @fabriciofx[/z](https://www.0crat.com/u/fabriciofx)

**Comment by 0crat** [comment]
This pull request [#740](https://github.com/yegor256/cactoos/issues/740) is assigned to @neonailol[/z](https://www.0crat.com/u/neonailol), here is [why](https://www.0crat.com/footprint/C63314D6Z/842230447); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer

**Comment by neonailol** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/cactoos/issues/740#issuecomment-376401816))

@neonailol The user @neonailol[/z](https://www.0crat.com/u/neonailol) resigned from [#740](https://github.com/yegor256/cactoos/issues/740), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
Tasks refusal is discouraged, see [§6](http://www.zerocracy.com/policy.html#6): -15 point(s) just awarded to @neonailol[/z](https://www.0crat.com/u/neonailol)

File: src/main/java/org/cactoos/io/Zip.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    /**
+     * Ctor.
+     *
+     * @param origin Origin directory.
+     */
+    public Zip(final Directory origin) {
+        this.origin = origin;
+    }
+
+    @Override public InputStream stream() throws IOException {
```
Comment (llorllale) [inline src/main/java/org/cactoos/io/Zip.java]: @SharplEr please fix formatting here - `@Override` should be above `public`

File: src/main/java/org/cactoos/io/Zip.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        final ByteArrayOutputStream out = new ByteArrayOutputStream();
+        try (final ZipOutputStream zip = new ZipOutputStream(out)) {
+            Path root = null;
+            for (final Path path : this.origin) {
+                if (root == null) {
```
Comment (llorllale) [inline src/main/java/org/cactoos/io/Zip.java]: @SharplEr do we need this null check?

File: src/test/java/org/cactoos/io/ZipTest.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ * @version $Id$
+ * @since 0.29
+ * @checkstyle JavadocMethodCheck (500 lines)
+ */
+public final class ZipTest {
+    /**
+     * Temporary folder.
+     */
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
```
Comment (llorllale) [inline src/test/java/org/cactoos/io/ZipTest.java]: @SharplEr should be `final`

**Comment by llorllale** [comment]
@SharplEr see my comments above

File: src/main/java/org/cactoos/io/Zip.java (hunk 2 above)
Comment (SharplEr) [reply src/main/java/org/cactoos/io/Zip.java]: @llorllale now when you are asking... Nevermind, It was the wrong idea anyway:) I removed this code.

File: src/main/java/org/cactoos/io/Zip.java (hunk 1 above)
Comment (SharplEr) [reply src/main/java/org/cactoos/io/Zip.java]: @llorllale  Done.

File: src/main/java/org/cactoos/io/Zip.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                final ZipEntry entry = new ZipEntry(
+                    root.relativize(path).toFile().getPath()
+                );
+                zip.putNextEntry(entry);
+                if (file.isFile()) {
+                    try (
+                        final FileInputStream input = new FileInputStream(file)
+                    ) {
+                        // @checkstyle MagicNumber (1 line)
+                        final byte[] buffer = new byte[1024];
```
[the remark this answers is not in the document]
Comment (SharplEr) [reply src/main/java/org/cactoos/io/Zip.java]: @llorllale Done.

File: src/test/java/org/cactoos/io/ZipTest.java (hunk 3 above)
Comment (SharplEr) [reply src/test/java/org/cactoos/io/ZipTest.java]: @llorllale Done.

**Comment by SharplEr** [comment]
@llorllale Done.

**Comment by llorllale** [comment]
@rultor merge

**Pull request merged**